### PR TITLE
enhance: Workspace Sync checks for changes before pushing & better handling when there is no upstream

### PR DESCRIPTION
**Pull request by SeriousBug**
We now check if there are any changes to be pushed before pushing. This way, if the user adds a vault they can't push to, sync will continue to function without any configuration as long as they don't start making changes to the repository.

We now also check if there is an upstream branch configured for the current branch, in addition to checking if there are remotes. This can easily happen if the user checks out a new branch. If it does happen, we now continue syncing other vaults, and notify the user that we can't pull or push that repository because it has no upstream branch.

### CODE REVIEW

File: packages/engine-server/src/topics/git.ts [hunk 1]
```diff
@@ -135,4 +135,41 @@ export class Git {
...
+      );
+      return _.trim(stdout);
+    } catch {
+      return undefined;
+    }
+  }
+
+  /**
+   *
+   * @param old: The old identifier (e.g. commit, tag, branch) that we are diffing against.
```
Comment (kevinslin) [inline packages/engine-server/src/topics/git.ts:153]: can the param name match the variable name? `eg. oldCommit`

File: packages/engine-server/src/topics/git.ts [hunk 2]
```diff
@@ -135,4 +135,41 @@ export class Git {
...
+   * @param new: The new identifier (e.g. commit, tag, branch) that we are diffing from.
```
Comment (kevinslin) [inline packages/engine-server/src/topics/git.ts:154]: see above

**Review by kevinslin — COMMENTED** [review]
minor doc string updates. do you know why the plugin test is failing?

**Comment by SeriousBug** [comment]
>  do you know why the plugin test is failing?

Not sure, but seems to be passing now.

**Review by kevinslin — APPROVED** [review]

**Pull request merged**